// remult/remult-mongo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MongoDataProvider = void 0;
var tslib_1 = require("tslib");
var mongodb_1 = require("mongodb");
var index_js_1 = require("./index.js");
var filter_consumer_bridge_to_sql_request_js_1 = require("./src/filter/filter-consumer-bridge-to-sql-request.js");
var remult_proxy_js_1 = require("./src/remult-proxy.js");
var RepositoryImplementation_js_1 = require("./src/remult3/RepositoryImplementation.js");
var repository_internals_js_1 = require("./src/remult3/repository-internals.js");
var sql_database_js_1 = require("./src/data-providers/sql-database.js");
var MongoDataProvider = /** @class */ (function () {
    function MongoDataProvider(db, client, options) {
        this.db = db;
        this.client = client;
        this.disableTransactions = false;
        this.session = options === null || options === void 0 ? void 0 : options.session;
        this.disableTransactions = Boolean(options === null || options === void 0 ? void 0 : options.disableTransactions);
    }
    MongoDataProvider.getDb = function (dataProvider) {
        var r = (dataProvider || remult_proxy_js_1.remult.dataProvider);
        if (!r.db)
            throw 'the data provider is not a MongoDataProvider';
        return { db: r.db, session: r.session };
    };
    MongoDataProvider.prototype.getEntityDataProvider = function (entity) {
        return new MongoEntityDataProvider(this.db, entity, this.session);
    };
    MongoDataProvider.prototype.transaction = function (action) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var session, db, err_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.disableTransactions) return [3 /*break*/, 2];
                        return [4 /*yield*/, action(this)];
                    case 1:
                        _a.sent();
                        return [3 /*break*/, 11];
                    case 2:
                        if (!this.client)
                            throw new Error("Can't use transactions within transactions");
                        return [4 /*yield*/, this.client.startSession()];
                    case 3:
                        session = _a.sent();
                        session.startTransaction();
                        db = this.client.db(this.db.databaseName);
                        _a.label = 4;
                    case 4:
                        _a.trys.push([4, 7, 9, 11]);
                        return [4 /*yield*/, action(new MongoDataProvider(db, undefined, { session: session }))];
                    case 5:
                        _a.sent();
                        return [4 /*yield*/, session.commitTransaction()];
                    case 6:
                        _a.sent();
                        return [3 /*break*/, 11];
                    case 7:
                        err_1 = _a.sent();
                        return [4 /*yield*/, session.abortTransaction()];
                    case 8:
                        _a.sent();
                        throw err_1;
                    case 9: return [4 /*yield*/, session.endSession()];
                    case 10:
                        _a.sent();
                        return [7 /*endfinally*/];
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    MongoDataProvider.filterToRaw = function (entity, condition) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var repo, b, _a, r;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        repo = (0, RepositoryImplementation_js_1.getRepository)(entity);
                        _a = FilterConsumerBridgeToMongo.bind;
                        return [4 /*yield*/, (0, filter_consumer_bridge_to_sql_request_js_1.dbNamesOfWithForceSqlExpression)(repo.metadata)];
                    case 1:
                        b = new (_a.apply(FilterConsumerBridgeToMongo, [void 0, _b.sent()]))();
                        b._addWhere = false;
                        return [4 /*yield*/, (0, repository_internals_js_1.getRepositoryInternals)(repo)._translateWhereToFilter(condition)];
                    case 2: return [4 /*yield*/, (_b.sent()).__applyToConsumer(b)];
                    case 3:
                        _b.sent();
                        return [4 /*yield*/, b.resolveWhere()];
                    case 4:
                        r = _b.sent();
                        return [2 /*return*/, r];
                }
            });
        });
    };
    return MongoDataProvider;
}());
exports.MongoDataProvider = MongoDataProvider;
var NULL = { $null: '$null' };
function isNull(x) {
    return (x === null || x === void 0 ? void 0 : x.$null) === NULL.$null;
}
var MongoEntityDataProvider = /** @class */ (function () {
    function MongoEntityDataProvider(db, entity, session) {
        this.db = db;
        this.entity = entity;
        this.session = session;
    }
    MongoEntityDataProvider.prototype.translateFromDb = function (row, nameProvider) {
        var e_1, _a;
        var result = {};
        try {
            for (var _b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var col = _c.value;
                var val = row[nameProvider.$dbNameOf(col)];
                if (isNull(val))
                    val = null;
                result[col.key] = fromDb(col, val);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return result;
    };
    MongoEntityDataProvider.prototype.translateToDb = function (row, nameProvider) {
        var e_2, _a;
        var result = {};
        try {
            for (var _b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var col = _c.value;
                var val = toDb(col, row[col.key]);
                if (val === null)
                    val = NULL;
                result[nameProvider.$dbNameOf(col)] = val;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return result;
    };
    MongoEntityDataProvider.prototype.count = function (where) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, collection, e, x, w;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.collection()];
                    case 1:
                        _a = _b.sent(), collection = _a.collection, e = _a.e;
                        x = new FilterConsumerBridgeToMongo(e);
                        where.__applyToConsumer(x);
                        return [4 /*yield*/, x.resolveWhere()];
                    case 2:
                        w = _b.sent();
                        return [4 /*yield*/, collection.countDocuments(w, { session: this.session })];
                    case 3: return [2 /*return*/, _b.sent()];
                }
            });
        });
    };
    MongoEntityDataProvider.prototype.find = function (options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, collection, e, x, where, op, _b, _c, s, _d, _e;
            var e_3, _f;
            var _this = this;
            return tslib_1.__generator(this, function (_g) {
                switch (_g.label) {
                    case 0: return [4 /*yield*/, this.collection()];
                    case 1:
                        _a = _g.sent(), collection = _a.collection, e = _a.e;
                        x = new FilterConsumerBridgeToMongo(e);
                        if (options === null || options === void 0 ? void 0 : options.where)
                            options.where.__applyToConsumer(x);
                        return [4 /*yield*/, x.resolveWhere()];
                    case 2:
                        where = _g.sent();
                        op = {
                            session: this.session,
                        };
                        if (options.limit) {
                            op.limit = options.limit;
                            if (options.page) {
                                op.skip = (options.page - 1) * options.limit;
                            }
                        }
                        if (options.orderBy) {
                            op.sort = {};
                            try {
                                for (_b = tslib_1.__values(options.orderBy.Segments), _c = _b.next(); !_c.done; _c = _b.next()) {
                                    s = _c.value;
                                    op.sort[e.$dbNameOf(s.field)] = s.isDescending ? -1 : 1;
                                }
                            }
                            catch (e_3_1) { e_3 = { error: e_3_1 }; }
                            finally {
                                try {
                                    if (_c && !_c.done && (_f = _b.return)) _f.call(_b);
                                }
                                finally { if (e_3) throw e_3.error; }
                            }
                        }
                        _e = (_d = Promise).all;
                        return [4 /*yield*/, collection
                                .find(where, op)
                                .map(function (x) { return _this.translateFromDb(x, e); })
                                .toArray()];
                    case 3: return [4 /*yield*/, _e.apply(_d, [_g.sent()])];
                    case 4: return [2 /*return*/, _g.sent()];
                }
            });
        });
    };
    MongoEntityDataProvider.prototype.update = function (id, data) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, collection, e, f, newR, keys, _b, _c, f_1, r, _d, _e;
            var e_4, _f;
            return tslib_1.__generator(this, function (_g) {
                switch (_g.label) {
                    case 0: return [4 /*yield*/, this.collection()];
                    case 1:
                        _a = _g.sent(), collection = _a.collection, e = _a.e;
                        f = new FilterConsumerBridgeToMongo(e);
                        index_js_1.Filter.fromEntityFilter(this.entity, this.entity.idMetadata.getIdFilter(id)).__applyToConsumer(f);
                        newR = {};
                        keys = Object.keys(data);
                        try {
                            for (_b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                                f_1 = _c.value;
                                if (!f_1.dbReadOnly && !f_1.isServerExpression) {
                                    if (keys.includes(f_1.key)) {
                                        newR[f_1.key] = toDb(f_1, data[f_1.key]);
                                    }
                                }
                            }
                        }
                        catch (e_4_1) { e_4 = { error: e_4_1 }; }
                        finally {
                            try {
                                if (_c && !_c.done && (_f = _b.return)) _f.call(_b);
                            }
                            finally { if (e_4) throw e_4.error; }
                        }
                        _e = (_d = collection).updateOne;
                        return [4 /*yield*/, f.resolveWhere()];
                    case 2: return [4 /*yield*/, _e.apply(_d, [_g.sent(), {
                                $set: newR,
                            },
                            { session: this.session }])];
                    case 3:
                        r = _g.sent();
                        return [2 /*return*/, (0, sql_database_js_1.getRowAfterUpdate)(this.entity, this, data, id, 'update')];
                }
            });
        });
    };
    MongoEntityDataProvider.prototype.delete = function (id) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, e, collection, f, _b, _c;
            return tslib_1.__generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, this.collection()];
                    case 1:
                        _a = _d.sent(), e = _a.e, collection = _a.collection;
                        f = new FilterConsumerBridgeToMongo(e);
                        index_js_1.Filter.fromEntityFilter(this.entity, this.entity.idMetadata.getIdFilter(id)).__applyToConsumer(f);
                        _c = (_b = collection).deleteOne;
                        return [4 /*yield*/, f.resolveWhere()];
                    case 2:
                        _c.apply(_b, [_d.sent(), { session: this.session }]);
                        return [2 /*return*/];
                }
            });
        });
    };
    MongoEntityDataProvider.prototype.insert = function (data) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, collection, e, r, _b, _c, _d;
            return tslib_1.__generator(this, function (_e) {
                switch (_e.label) {
                    case 0: return [4 /*yield*/, this.collection()];
                    case 1:
                        _a = _e.sent(), collection = _a.collection, e = _a.e;
                        _c = (_b = collection).insertOne;
                        return [4 /*yield*/, this.translateToDb(data, e)];
                    case 2: return [4 /*yield*/, _c.apply(_b, [_e.sent(), {
                                session: this.session,
                            }])];
                    case 3:
                        r = _e.sent();
                        _d = this.translateFromDb;
                        return [4 /*yield*/, collection.findOne({ _id: r.insertedId }, { session: this.session })];
                    case 4: return [4 /*yield*/, _d.apply(this, [_e.sent(), e])];
                    case 5: return [2 /*return*/, _e.sent()];
                }
            });
        });
    };
    MongoEntityDataProvider.prototype.collection = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var e, collection;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, filter_consumer_bridge_to_sql_request_js_1.dbNamesOf)(this.entity)];
                    case 1:
                        e = _a.sent();
                        collection = this.db.collection(e.$entityName);
                        return [2 /*return*/, { e: e, collection: collection }];
                }
            });
        });
    };
    return MongoEntityDataProvider;
}());
var FilterConsumerBridgeToMongo = /** @class */ (function () {
    function FilterConsumerBridgeToMongo(nameProvider) {
        this.nameProvider = nameProvider;
        this._addWhere = true;
        this.promises = [];
        this.result = [];
    }
    FilterConsumerBridgeToMongo.prototype.resolveWhere = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var p, p_1, p_1_1, pr, e_5_1;
            var e_5, _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(this.promises.length > 0)) return [3 /*break*/, 9];
                        p = this.promises;
                        this.promises = [];
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 6, 7, 8]);
                        p_1 = (e_5 = void 0, tslib_1.__values(p)), p_1_1 = p_1.next();
                        _b.label = 2;
                    case 2:
                        if (!!p_1_1.done) return [3 /*break*/, 5];
                        pr = p_1_1.value;
                        return [4 /*yield*/, pr];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4:
                        p_1_1 = p_1.next();
                        return [3 /*break*/, 2];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_5_1 = _b.sent();
                        e_5 = { error: e_5_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (p_1_1 && !p_1_1.done && (_a = p_1.return)) _a.call(p_1);
                        }
                        finally { if (e_5) throw e_5.error; }
                        return [7 /*endfinally*/];
                    case 8: return [3 /*break*/, 0];
                    case 9:
                        if (this.result.length > 0)
                            return [2 /*return*/, { $and: this.result.map(function (x) { return x(); }) }];
                        else
                            return [2 /*return*/, {}];
                        return [2 /*return*/];
                }
            });
        });
    };
    FilterConsumerBridgeToMongo.prototype.custom = function (key, customItem) {
        throw new Error('Custom filter should be translated before it gets here');
    };
    FilterConsumerBridgeToMongo.prototype.or = function (orElements) {
        var _this = this;
        this.promises.push((function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var result, orElements_1, orElements_1_1, element, f, where, e_6_1;
            var e_6, _a;
            var _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        result = [];
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 6, 7, 8]);
                        orElements_1 = tslib_1.__values(orElements), orElements_1_1 = orElements_1.next();
                        _c.label = 2;
                    case 2:
                        if (!!orElements_1_1.done) return [3 /*break*/, 5];
                        element = orElements_1_1.value;
                        f = new FilterConsumerBridgeToMongo(this.nameProvider);
                        f._addWhere = false;
                        element.__applyToConsumer(f);
                        return [4 /*yield*/, f.resolveWhere()];
                    case 3:
                        where = _c.sent();
                        if ((_b = where === null || where === void 0 ? void 0 : where.$and) === null || _b === void 0 ? void 0 : _b.length) {
                            result.push(where);
                        }
                        else
                            return [2 /*return*/]; //since empty or is all rows;
                        _c.label = 4;
                    case 4:
                        orElements_1_1 = orElements_1.next();
                        return [3 /*break*/, 2];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_6_1 = _c.sent();
                        e_6 = { error: e_6_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (orElements_1_1 && !orElements_1_1.done && (_a = orElements_1.return)) _a.call(orElements_1);
                        }
                        finally { if (e_6) throw e_6.error; }
                        return [7 /*endfinally*/];
                    case 8:
                        this.result.push(function () { return ({
                            $or: result,
                        }); });
                        return [2 /*return*/];
                }
            });
        }); })());
    };
    FilterConsumerBridgeToMongo.prototype.not = function (element) {
        var _this = this;
        this.promises.push((function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var f, where;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        f = new FilterConsumerBridgeToMongo(this.nameProvider);
                        f._addWhere = false;
                        element.__applyToConsumer(f);
                        return [4 /*yield*/, f.resolveWhere()];
                    case 1:
                        where = _a.sent();
                        if (where) {
                            this.result.push(function () { return ({
                                $nor: [where],
                            }); });
                        }
                        return [2 /*return*/];
                }
            });
        }); })());
    };
    FilterConsumerBridgeToMongo.prototype.isNull = function (col) {
        this.add(col, NULL, '$eq');
    };
    FilterConsumerBridgeToMongo.prototype.isNotNull = function (col) {
        this.add(col, NULL, '$ne');
    };
    FilterConsumerBridgeToMongo.prototype.isIn = function (col, val) {
        var _this = this;
        this.result.push(function () {
            var _a;
            return (_a = {},
                _a[_this.nameProvider.$dbNameOf(col)] = {
                    $in: val.map(function (x) { return toDb(col, x); }),
                },
                _a);
        });
    };
    FilterConsumerBridgeToMongo.prototype.isEqualTo = function (col, val) {
        this.add(col, val, '$eq');
    };
    FilterConsumerBridgeToMongo.prototype.isDifferentFrom = function (col, val) {
        this.add(col, val, '$ne');
    };
    FilterConsumerBridgeToMongo.prototype.isGreaterOrEqualTo = function (col, val) {
        this.add(col, val, '$gte');
    };
    FilterConsumerBridgeToMongo.prototype.isGreaterThan = function (col, val) {
        this.add(col, val, '$gt');
    };
    FilterConsumerBridgeToMongo.prototype.isLessOrEqualTo = function (col, val) {
        this.add(col, val, '$lte');
    };
    FilterConsumerBridgeToMongo.prototype.isLessThan = function (col, val) {
        this.add(col, val, '$lt');
    };
    FilterConsumerBridgeToMongo.prototype.containsCaseInsensitive = function (col, val) {
        this.add(col, val, '$regex', { $options: 'i' });
    };
    FilterConsumerBridgeToMongo.prototype.notContainsCaseInsensitive = function (col, val) {
        var _this = this;
        this.result.push(function () {
            var _a;
            return (_a = {},
                _a[_this.nameProvider.$dbNameOf(col)] = {
                    $not: {
                        $regex: isNull(val) ? val : toDb(col, val),
                        $options: 'i',
                    },
                },
                _a);
        });
    };
    FilterConsumerBridgeToMongo.prototype.startsWithCaseInsensitive = function (col, val) {
        this.add(col, "^".concat(val), '$regex', { $options: 'i' });
    };
    FilterConsumerBridgeToMongo.prototype.endsWithCaseInsensitive = function (col, val) {
        this.add(col, "".concat(val, "$"), '$regex', { $options: 'i' });
    };
    FilterConsumerBridgeToMongo.prototype.add = function (col, val, operator, moreOptions) {
        var _this = this;
        this.result.push(function () {
            var _a, _b;
            return (_a = {},
                _a[_this.nameProvider.$dbNameOf(col)] = tslib_1.__assign((_b = {}, _b[operator] = isNull(val) ? val : toDb(col, val), _b), moreOptions),
                _a);
        });
    };
    FilterConsumerBridgeToMongo.prototype.databaseCustom = function (databaseCustom) {
        throw 'error';
        //   this.promises.push((async () => {
        //     if (databaseCustom?.buildSql) {
        //       let item = new CustomSqlFilterBuilder(this.knex);
        //       await databaseCustom.buildSql(item);
        //       if (item.sql) {
        //         this.addToWhere("(" + item.sql + ")");
        //       }
        //     }
        //   })());
    };
    return FilterConsumerBridgeToMongo;
}());
function toDb(col, val) {
    if (col.valueConverter.fieldTypeInDb == 'dbid')
        return val ? new mongodb_1.ObjectId(val) : val === null ? null : undefined;
    return col.valueConverter.toDb(val);
}
function fromDb(col, val) {
    if (col.valueConverter.fieldTypeInDb == 'dbid')
        return val ? val.toHexString() : val === null ? null : undefined;
    return col.valueConverter.fromDb(val);
}
