// remult/remult-sveltekit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.remultSveltekit = void 0;
var tslib_1 = require("tslib");
var index_js_1 = require("./server/index.js");
function remultSveltekit(options) {
    var _this = this;
    var result = (0, index_js_1.createRemultServer)(options, {
        buildGenericRequestInfo: function (event) { return ({
            url: event.request.url,
            method: event.request.method,
            on: function (e, do1) {
                if (e === 'close') {
                    event.locals['_tempOnClose'] = do1;
                }
            },
        }); },
        getRequestBody: function (event) { return event.request.json(); },
    });
    var serverHandler = function (event) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
        var sseResponse, response, responseFromRemultHandler, res;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    sseResponse = undefined;
                    event.locals['_tempOnClose'] = function () { };
                    response = {
                        end: function () { },
                        json: function () { },
                        send: function () { },
                        status: function () {
                            return response;
                        },
                        write: function () { },
                        writeHead: function (status, headers) {
                            if (status === 200 && headers) {
                                var contentType = headers['Content-Type'];
                                if (contentType === 'text/event-stream') {
                                    var messages_1 = [];
                                    response.write = function (x) { return messages_1.push(x); };
                                    var stream = new ReadableStream({
                                        start: function (controller) {
                                            var e_1, _a;
                                            try {
                                                for (var messages_2 = tslib_1.__values(messages_1), messages_2_1 = messages_2.next(); !messages_2_1.done; messages_2_1 = messages_2.next()) {
                                                    var message = messages_2_1.value;
                                                    controller.enqueue(message);
                                                }
                                            }
                                            catch (e_1_1) { e_1 = { error: e_1_1 }; }
                                            finally {
                                                try {
                                                    if (messages_2_1 && !messages_2_1.done && (_a = messages_2.return)) _a.call(messages_2);
                                                }
                                                finally { if (e_1) throw e_1.error; }
                                            }
                                            response.write = function (data) {
                                                controller.enqueue(data);
                                            };
                                        },
                                        cancel: function () {
                                            response.write = function () { };
                                            event.locals['_tempOnClose']();
                                        },
                                    });
                                    sseResponse = new Response(stream, { headers: headers });
                                }
                            }
                        },
                    };
                    return [4 /*yield*/, result.handle(event, response)];
                case 1:
                    responseFromRemultHandler = _a.sent();
                    if (sseResponse !== undefined) {
                        return [2 /*return*/, sseResponse];
                    }
                    if (responseFromRemultHandler) {
                        if (responseFromRemultHandler.html)
                            return [2 /*return*/, new Response(responseFromRemultHandler.html, {
                                    status: responseFromRemultHandler.statusCode,
                                    headers: {
                                        'Content-Type': 'text/html',
                                    },
                                })];
                        res = new Response(JSON.stringify(responseFromRemultHandler.data), {
                            status: responseFromRemultHandler.statusCode,
                        });
                        return [2 /*return*/, res];
                    }
                    return [2 /*return*/, new Response('Not Found', {
                            status: 404,
                        })];
            }
        });
    }); };
    var handler = function (_a) {
        var event = _a.event, resolve = _a.resolve;
        return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var result_1;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!event.url.pathname.startsWith(options.rootPath)) return [3 /*break*/, 2];
                        return [4 /*yield*/, serverHandler(event)];
                    case 1:
                        result_1 = _b.sent();
                        if (result_1 != null && (result_1 === null || result_1 === void 0 ? void 0 : result_1.status) != 404)
                            return [2 /*return*/, result_1];
                        _b.label = 2;
                    case 2: return [2 /*return*/, new Promise(function (res) {
                            result.withRemult(event, undefined, function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var _a;
                                return tslib_1.__generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            _a = res;
                                            return [4 /*yield*/, resolve(event)];
                                        case 1:
                                            _a.apply(void 0, [_b.sent()]);
                                            return [2 /*return*/];
                                    }
                                });
                            }); });
                        })];
                }
            });
        });
    };
    return Object.assign(handler, {
        getRemult: function (req) { return result.getRemult(req); },
        openApiDoc: function (options) { return result.openApiDoc(options); },
        withRemult: function (request, what) {
            return result.withRemultAsync(request, what);
        },
        GET: serverHandler,
        PUT: serverHandler,
        POST: serverHandler,
        DELETE: serverHandler,
    });
}
exports.remultSveltekit = remultSveltekit;
