// remult/esm/src/remult3/Fields.js
import { createId } from '@paralleldrive/cuid2';
import { v4 as uuid } from 'uuid';
import { ValueConverters } from '../valueConverters.js';
import { buildOptions, fieldOptionalValuesFunctionKey, } from './RepositoryImplementation.js';
import { Validators, createValueValidator, getEnumValues, } from '../validators.js';
import { relationInfoMemberInOptions } from './relationInfoMember.js';
import { remultStatic } from '../remult-static.js';
import { addValidator } from './addValidator.js';
const validateNumber = createValueValidator((x) => {
    return !isNaN(x) && isFinite(x);
});
export class Fields {
    /**
     * Stored as a JSON.stringify - to store as json use Fields.json
     */
    static object(...options) {
        return Field(undefined, ...options);
    }
    static json(...options) {
        let op = options;
        if (op.valueConverter && !op.valueConverter.fieldTypeInDb)
            //@ts-ignore
            op.valueConverter.fieldTypeInDb = 'json';
        return Field(undefined, {
            valueConverter: {
                fieldTypeInDb: 'json',
            },
        }, ...options);
    }
    static dateOnly(...options) {
        return Field(() => Date, {
            valueConverter: ValueConverters.DateOnly,
        }, ...options);
    }
    static date(...options) {
        return Field(() => Date, ...options);
    }
    static integer(...options) {
        return Field(() => Number, {
            valueConverter: ValueConverters.Integer,
            validate: validateNumber,
        }, ...options);
    }
    static autoIncrement(...options) {
        return Field(() => Number, {
            allowApiUpdate: false,
            dbReadOnly: true,
            valueConverter: {
                ...ValueConverters.Integer,
                fieldTypeInDb: 'autoincrement',
            },
        }, ...options);
    }
    static number(...options) {
        return Field(() => Number, {
            validate: validateNumber,
        }, ...options);
    }
    static createdAt(...options) {
        return Field(() => Date, {
            allowApiUpdate: false,
            saving: (_, ref, { isNew }) => {
                if (isNew)
                    ref.value = new Date();
            },
        }, ...options);
    }
    static updatedAt(...options) {
        return Field(() => Date, {
            allowApiUpdate: false,
            saving: (_, ref) => {
                ref.value = new Date();
            },
        }, ...options);
    }
    static uuid(...options) {
        return Field(() => String, {
            allowApiUpdate: false,
            defaultValue: () => uuid(),
            saving: (_, r) => {
                if (!r.value)
                    r.value = uuid();
            },
        }, ...options);
    }
    /**
     * A CUID (Collision Resistant Unique Identifier) field.
     * This id value is determined on the backend on insert, and can't be updated through the API.
     * The CUID is generated using the `@paralleldrive/cuid2` npm package.
     */
    static cuid(...options) {
        return Field(() => String, {
            allowApiUpdate: false,
            defaultValue: () => createId(),
            saving: (_, r) => {
                if (!r.value)
                    r.value = createId();
            },
        }, ...options);
    }
    // static id<entityType = any>(
    //   options?: FieldOptions<entityType, string> & { idFactory?: () => string },
    // ): ClassFieldDecorator<entityType, string | undefined> {
    //   let buildId = options?.idFactory ?? createId
    //   return Field(() => String, {
    //     allowApiUpdate: false,
    //     defaultValue: () => buildId(),
    //     saving: (_, r) => {
    //       if (!r.value) r.value = buildId()
    //     },
    //   })
    // }
    /**
   * Defines a field that can hold a value from a specified set of string literals.
   * @param {() => readonly valueType[]} optionalValues - A function that returns an array of allowed string literals.
   * @returns {ClassFieldDecorator<entityType, valueType | undefined>} - A class field decorator.
   *
   * @example
   
   * class MyEntity {
   *   .@Fields.literal(() => ['open', 'closed', 'frozen', 'in progress'] as const)
   *   status: 'open' | 'closed' | 'frozen' | 'in progress' = 'open';
   * }
   
   *
   * // This defines a field `status` in `MyEntity` that can only hold the values 'open', 'closed', 'frozen', or 'in progress'.
   *
   * @example
   * // For better reusability and maintainability:
   
   * const statuses = ['open', 'closed', 'frozen', 'in progress'] as const;
   * type StatusType = typeof statuses[number];
   *
   * class MyEntity {
   *   .@Fields.literal(() => statuses)
   *   status: StatusType = 'open';
   * }
   
   *
   * // This approach allows easy management and updates of the allowed values for the `status` field.
   */
    static literal(optionalValues, ...options) {
        return Fields.string({
            validate: (entity, event) => Validators.in(optionalValues())(entity, event),
            //@ts-expect-error as we are adding this to options without it being defined in options
            [fieldOptionalValuesFunctionKey]: optionalValues,
        }, ...options);
    }
    static enum(enumType, ...options) {
        let valueConverter;
        return Field(() => 
        //@ts-ignore
        enumType(), {
            validate: (entity, event) => Validators.enum(enumType())(entity, event),
            [fieldOptionalValuesFunctionKey]: () => getEnumValues(enumType()),
        }, ...options, (options) => {
            options[fieldOptionalValuesFunctionKey] = () => getEnumValues(enumType());
            if (valueConverter === undefined) {
                let enumObj = enumType();
                let enumValues = getEnumValues(enumObj);
                valueConverter = enumValues.find((x) => typeof x === 'string')
                    ? ValueConverters.String
                    : ValueConverters.Integer;
            }
            if (!options.valueConverter) {
                options.valueConverter = valueConverter;
            }
            else if (!options.valueConverter.fieldTypeInDb) {
                //@ts-ignore
                options.valueConverter.fieldTypeInDb = valueConverter.fieldTypeInDb;
            }
        });
    }
    static string(...options) {
        return Field(() => String, ...options);
    }
    static boolean(...options) {
        return Field(() => Boolean, ...options);
    }
}
export class Relations {
    /**
     * Define a to-one relation between entities, indicating a one-to-one relationship.
     * If no field or fields are provided, it will automatically create a field in the database
     * to represent the relation.
     *
     * @param toEntityType A function that returns the target entity type.
     * @param options (Optional): An object containing options for configuring the to-one relation.
     * @returns A decorator function to apply the to-one relation to an entity field.
     *
     * Example usage:
     * ```
     * @Relations.toOne(() => Customer)
     * customer?: Customer;
     * ```
     * ```
     * Fields.string()
     * customerId?: string;
     *
     * @Relations.toOne(() => Customer, "customerId")
     * customer?: Customer;
     * ```
     * ```
     * Fields.string()
     * customerId?: string;
     *
     * @Relations.toOne(() => Customer, {
     *   field: "customerId",
     *   defaultIncluded: true
     * })
     * customer?: Customer;
     * ```
     * ```
     * Fields.string()
     * customerId?: string;
     *
     * @Relations.toOne(() => Customer, {
     *   fields: {
     *     customerId: "id",
     *   },
     * })
     * customer?: Customer;
     * ```
     */
    static toOne(toEntityType, options) {
        let op = (typeof options === 'string'
            ? { field: options }
            : !options
                ? {}
                : options);
        if (!op.field && !op.fields && !op.findOptions)
            //@ts-ignore
            return Field(toEntityType, {
                ...op,
                ...relationInfoMemberInOptions(toEntityType, 'reference'),
            });
        return Field(() => undefined, {
            ...op,
            serverExpression: () => undefined,
            ...relationInfoMemberInOptions(toEntityType, 'toOne'),
        });
    }
    static toMany(toEntityType, options) {
        let op = (typeof options === 'string'
            ? { field: options }
            : options);
        return Field(() => undefined, {
            ...op,
            serverExpression: () => undefined,
            ...relationInfoMemberInOptions(toEntityType, 'toMany'),
        });
    }
}
/**Decorates fields that should be used as fields.
 * for more info see: [Field Types](https://remult.dev/docs/field-types.html)
 *
 * FieldOptions can be set in two ways:
 * @example
 * // as an object
 * @Fields.string({ includeInApi:false })
 * title='';
 * @example
 * // as an arrow function that receives `remult` as a parameter
 * @Fields.string((options,remult) => options.includeInApi = true)
 * title='';
 */
export function Field(valueType, ...options) {
    // import ANT!!!! if you call this in another decorator, make sure to set It's return type correctly with the | undefined
    return (target, context, c) => {
        const key = typeof context === 'string' ? context : context.name.toString();
        let factory = (remult) => {
            let r = buildOptions(options, remult);
            if (r.required) {
                r.validate = addValidator(r.validate, Validators.required, true);
            }
            if (r.maxLength) {
                r.validate = addValidator(r.validate, Validators.maxLength(r.maxLength));
            }
            if (r.minLength) {
                r.validate = addValidator(r.validate, Validators.minLength(r.minLength));
            }
            if (!r.valueType && valueType) {
                r.valueType = valueType();
            }
            if (!r.key) {
                r.key = key;
            }
            if (!r.dbName)
                r.dbName = r.key;
            let type = r.valueType;
            if (!type) {
                type =
                    typeof Reflect.getMetadata == 'function'
                        ? Reflect.getMetadata('design:type', target, key)
                        : [];
                r.valueType = type;
            }
            if (!r.target)
                r.target = target;
            return r;
        };
        checkTarget(target);
        let names = remultStatic.columnsOfType.get(target.constructor);
        if (!names) {
            names = [];
            remultStatic.columnsOfType.set(target.constructor, names);
        }
        let set = names.find((x) => x.key == key);
        if (!set)
            names.push({
                key,
                settings: factory,
            });
        else {
            let prev = set.settings;
            set.settings = (c) => {
                let prevO = prev(c);
                let curr = factory(c);
                return Object.assign(prevO, curr);
            };
        }
    };
}
export function checkTarget(target) {
    if (!target)
        throw new Error("Set the 'experimentalDecorators:true' option in your 'tsconfig' or 'jsconfig' (target undefined)");
}
