// remult/src/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnumValues = exports.createValidatorWithArgs = exports.createValueValidatorWithArgs = exports.createValueValidator = exports.valueValidator = exports.createValidator = exports.Validators = void 0;
var tslib_1 = require("tslib");
var Validators = /** @class */ (function () {
    function Validators() {
    }
    var _a;
    _a = Validators;
    Validators.required = createValidator(function (_, e) { return tslib_1.__awaiter(void 0, void 0, void 0, function () { return tslib_1.__generator(_a, function (_b) {
        return [2 /*return*/, e.value != null && e.value != undefined && e.value !== ''];
    }); }); }, 'Should not be empty');
    Validators.unique = createValidator(function (_, e) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var _b;
        return tslib_1.__generator(_a, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!e.entityRef)
                        throw 'unique validation may only work on columns that are attached to an entity';
                    if (!(e.isBackend() && (e.isNew || e.valueChanged()))) return [3 /*break*/, 2];
                    return [4 /*yield*/, e.entityRef.repository.count((_b = {},
                            _b[e.metadata.key] = e.value,
                            _b))];
                case 1: return [2 /*return*/, ((_c.sent()) == 0)];
                case 2: return [2 /*return*/, true];
            }
        });
    }); }, 'already exists');
    /**
     * @deprecated use `unique` instead - it also runs only on the backend
     */
    Validators.uniqueOnBackend = createValidator(function (_, e) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var _b;
        return tslib_1.__generator(_a, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!(e.isBackend() && (e.isNew || e.valueChanged()))) return [3 /*break*/, 2];
                    return [4 /*yield*/, e.entityRef.repository.count((_b = {},
                            _b[e.metadata.key] = e.value,
                            _b))];
                case 1: return [2 /*return*/, ((_c.sent()) == 0)];
                case 2: return [2 /*return*/, true];
            }
        });
    }); }, _a.unique.defaultMessage);
    Validators.regex = createValueValidatorWithArgs(function (val, regex) {
        return regex.test(val);
    });
    Validators.email = createValueValidator(function (val) { return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(val); }, 'Invalid Email');
    Validators.url = createValueValidator(function (val) { return !!new URL(val); }, 'Invalid Url');
    Validators.in = createValueValidatorWithArgs(function (val, values) { return values.includes(val); }, function (values) {
        return "Value must be one of: ".concat(values
            .map(function (y) {
            return typeof y === 'object'
                ? y['id'] !== undefined
                    ? y['id']
                    : y.toString()
                : y;
        })
            .join(', '));
    });
    Validators.notNull = createValueValidator(function (val) { return val != null; }, 'Should not be null');
    Validators.enum = createValueValidatorWithArgs(function (value, enumObj) { return Object.values(enumObj).includes(value); }, function (enumObj) { return "Value must be one of ".concat(getEnumValues(enumObj).join(', ')); });
    Validators.relationExists = createValidator(function (_, e) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var _b;
        return tslib_1.__generator(_a, function (_c) {
            switch (_c.label) {
                case 0:
                    if (e.valueIsNull())
                        return [2 /*return*/, true];
                    if (!e.isBackend())
                        return [2 /*return*/, true];
                    _b = Boolean;
                    return [4 /*yield*/, e.load()];
                case 1: return [2 /*return*/, _b.apply(void 0, [_c.sent()])];
            }
        });
    }); }, 'Relation value does not exist');
    Validators.maxLength = createValueValidatorWithArgs(function (val, maxLength) { return val.length <= maxLength; }, function (maxLength) { return "Value must be at most ".concat(maxLength, " characters"); });
    Validators.minLength = createValueValidatorWithArgs(function (val, minLength) { return val.length >= minLength; }, function (maxLength) { return "Value must be at least ".concat(maxLength, " characters"); });
    Validators.defaultMessage = 'Invalid value';
    return Validators;
}());
exports.Validators = Validators;
function createValidator(validate, defaultMessage) {
    var _this = this;
    var validation = function (entity, e, message) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
        var valid;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, validate(entity, e)];
                case 1:
                    valid = _b.sent();
                    if (typeof valid === 'string' && valid.length > 0)
                        e.error = valid;
                    else if (!valid)
                        e.error =
                            (typeof message === 'function' && message(entity, e, undefined)) ||
                                message ||
                                (typeof defaultMessage === 'function' &&
                                    defaultMessage(entity, e, undefined)) ||
                                defaultMessage ||
                                Validators.defaultMessage;
                    return [2 /*return*/];
            }
        });
    }); };
    var result = function (entityOrMessage, e, message) {
        if (typeof entityOrMessage === 'string' ||
            entityOrMessage === 'function' ||
            (entityOrMessage === undefined && e === undefined)) {
            return function (entity, e, message) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, validation(entity, e, entityOrMessage || message)];
                    case 1: return [2 /*return*/, _b.sent()];
                }
            }); }); };
        }
        return validation(entityOrMessage, e, message);
    };
    Object.defineProperty(result, 'defaultMessage', {
        get: function () {
            return defaultMessage;
        },
        set: function (val) {
            defaultMessage = val;
        },
        enumerable: true,
    });
    //@ts-ignore
    return Object.assign(result, {
        withMessage: function (message) {
            return function (entity, e) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_b) {
                return [2 /*return*/, result(entity, e, message)];
            }); }); };
        },
    });
}
exports.createValidator = createValidator;
function valueValidator(validate, defaultMessage) {
    return function (entity, e) {
        return validate(e.value) || defaultMessage || false;
    };
}
exports.valueValidator = valueValidator;
function createValueValidator(validate, defaultMessage) {
    return createValidator(function (_, e) {
        if (e.value === undefined || e.value === null)
            return true;
        return validate(e.value);
    }, defaultMessage);
}
exports.createValueValidator = createValueValidator;
function createValueValidatorWithArgs(validate, defaultMessage) {
    var result = createValidatorWithArgsInternal(function (_, e, args) {
        if (e.value === undefined || e.value === null)
            return true;
        return validate(e.value, args);
    }, function (_, e, args) {
        return (typeof defaultMessage === 'function' && defaultMessage(args)) ||
            defaultMessage;
    }, true);
    return Object.assign(function (entity, e) { return result(entity, e); }, {
        get defaultMessage() {
            return defaultMessage;
        },
        set defaultMessage(val) {
            defaultMessage = val;
        },
    });
}
exports.createValueValidatorWithArgs = createValueValidatorWithArgs;
function createValidatorWithArgs(validate, defaultMessage) {
    return createValidatorWithArgsInternal(validate, defaultMessage);
}
exports.createValidatorWithArgs = createValidatorWithArgs;
function createValidatorWithArgsInternal(validate, defaultMessage, isValueValidator) {
    var _this = this;
    if (isValueValidator === void 0) { isValueValidator = false; }
    var result = function (args, message) {
        return function (entity, e) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var valid;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, validate(entity, e, args)];
                    case 1:
                        valid = _b.sent();
                        if (typeof valid === 'string')
                            e.error = valid;
                        else if (!valid)
                            e.error = message
                                ? typeof message === 'function'
                                    ? isValueValidator
                                        ? message(args)
                                        : message(entity, e, args)
                                    : message
                                : defaultMessage
                                    ? typeof defaultMessage === 'function'
                                        ? defaultMessage(entity, e, args)
                                        : defaultMessage
                                    : Validators.defaultMessage;
                        return [2 /*return*/];
                }
            });
        }); };
    };
    return Object.assign(result, {
        get defaultMessage() {
            return defaultMessage;
        },
        set defaultMessage(val) {
            defaultMessage = val;
        },
    });
}
function getEnumValues(enumObj) {
    return Object.values(enumObj).filter(function (x) { return typeof enumObj[x] !== 'number'; });
}
exports.getEnumValues = getEnumValues;
