// remult/src/data-providers/array-entity-data-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayEntityDataProvider = void 0;
var tslib_1 = require("tslib");
var CompoundIdField_js_1 = require("../CompoundIdField.js");
var filter_consumer_bridge_to_sql_request_js_1 = require("../filter/filter-consumer-bridge-to-sql-request.js");
var filter_interfaces_js_1 = require("../filter/filter-interfaces.js");
var ArrayEntityDataProvider = /** @class */ (function () {
    function ArrayEntityDataProvider(entity, rows) {
        this.entity = entity;
        this.rows = rows;
    }
    ArrayEntityDataProvider.rawFilter = function (filter) {
        var _a;
        return _a = {},
            _a[filter_interfaces_js_1.customDatabaseFilterToken] = {
                arrayFilter: filter,
            },
            _a;
    };
    //@internal
    ArrayEntityDataProvider.prototype.init = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, r;
            var e_1, _d;
            return tslib_1.__generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        if (this.__names)
                            return [2 /*return*/, this.__names];
                        _a = this;
                        return [4 /*yield*/, (0, filter_consumer_bridge_to_sql_request_js_1.dbNamesOf)(this.entity, function (x) { return x; })];
                    case 1:
                        _a.__names = _e.sent();
                        try {
                            for (_b = tslib_1.__values(this.rows()), _c = _b.next(); !_c.done; _c = _b.next()) {
                                r = _c.value;
                                this.verifyThatRowHasAllNotNullColumns(r, this.__names);
                            }
                        }
                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                        finally {
                            try {
                                if (_c && !_c.done && (_d = _b.return)) _d.call(_b);
                            }
                            finally { if (e_1) throw e_1.error; }
                        }
                        return [2 /*return*/, this.__names];
                }
            });
        });
    };
    //@internal
    ArrayEntityDataProvider.prototype.verifyThatRowHasAllNotNullColumns = function (r, names) {
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var f = _c.value;
                var key = names.$dbNameOf(f);
                if (!f.isServerExpression)
                    if (!f.allowNull) {
                        if (r[key] === undefined || r[key] === null) {
                            var val = undefined;
                            if (f.valueType === Boolean)
                                val = false;
                            else if (f.valueType === Number)
                                val = 0;
                            else if (f.valueType === String)
                                val = '';
                            r[key] = val;
                        }
                    }
                    else if (r[key] === undefined)
                        r[key] = null;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    ArrayEntityDataProvider.prototype.count = function (where) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var rows, names, j, i, x;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        rows = this.rows();
                        return [4 /*yield*/, this.init()];
                    case 1:
                        names = _a.sent();
                        j = 0;
                        for (i = 0; i < rows.length; i++) {
                            if (!where) {
                                j++;
                            }
                            else {
                                x = new FilterConsumerBridgeToObject(rows[i], names);
                                where.__applyToConsumer(x);
                                if (x.ok)
                                    j++;
                            }
                        }
                        return [2 /*return*/, j];
                }
            });
        });
    };
    ArrayEntityDataProvider.prototype.find = function (options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var rows, dbNames;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        rows = this.rows();
                        return [4 /*yield*/, this.init()];
                    case 1:
                        dbNames = _a.sent();
                        if (options) {
                            if (options.where) {
                                rows = rows.filter(function (i) {
                                    var x = new FilterConsumerBridgeToObject(i, dbNames);
                                    options.where.__applyToConsumer(x);
                                    return x.ok;
                                });
                            }
                            if (options.orderBy) {
                                rows = rows.sort(function (a, b) {
                                    return options.orderBy.compare(a, b, dbNames.$dbNameOf);
                                });
                            }
                            rows = pageArray(rows, options);
                        }
                        if (rows)
                            return [2 /*return*/, rows.map(function (i) {
                                    return _this.translateFromJson(i, dbNames);
                                })];
                        return [2 /*return*/, []];
                }
            });
        });
    };
    //@internal
    ArrayEntityDataProvider.prototype.translateFromJson = function (row, dbNames) {
        var e_3, _a;
        var result = {};
        try {
            for (var _b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var col = _c.value;
                result[col.key] = col.valueConverter.fromJson(row[dbNames.$dbNameOf(col)]);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return result;
    };
    //@internal
    ArrayEntityDataProvider.prototype.translateToJson = function (row, dbNames) {
        var e_4, _a;
        var result = {};
        try {
            for (var _b = tslib_1.__values(this.entity.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var col = _c.value;
                if (!(0, filter_consumer_bridge_to_sql_request_js_1.isDbReadonly)(col, dbNames))
                    result[dbNames.$dbNameOf(col)] = col.valueConverter.toJson(row[col.key]);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return result;
    };
    //@internal
    ArrayEntityDataProvider.prototype.idMatches = function (id, names) {
        var _this = this;
        return function (item) {
            var x = new FilterConsumerBridgeToObject(item, names);
            filter_interfaces_js_1.Filter.fromEntityFilter(_this.entity, _this.entity.idMetadata.getIdFilter(id)).__applyToConsumer(x);
            return x.ok;
        };
    };
    ArrayEntityDataProvider.prototype.update = function (id, data) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var names, idMatches, keys, i, r, newR, _a, _b, f;
            var e_5, _c;
            return tslib_1.__generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        names = _d.sent();
                        idMatches = this.idMatches(id, names);
                        keys = Object.keys(data);
                        for (i = 0; i < this.rows().length; i++) {
                            r = this.rows()[i];
                            if (idMatches(r)) {
                                newR = tslib_1.__assign({}, r);
                                try {
                                    for (_a = (e_5 = void 0, tslib_1.__values(this.entity.fields)), _b = _a.next(); !_b.done; _b = _a.next()) {
                                        f = _b.value;
                                        if (!(0, filter_consumer_bridge_to_sql_request_js_1.isDbReadonly)(f, names)) {
                                            if (keys.includes(f.key)) {
                                                newR[names.$dbNameOf(f)] = f.valueConverter.toJson(data[f.key]);
                                            }
                                        }
                                    }
                                }
                                catch (e_5_1) { e_5 = { error: e_5_1 }; }
                                finally {
                                    try {
                                        if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                                    }
                                    finally { if (e_5) throw e_5.error; }
                                }
                                this.verifyThatRowHasAllNotNullColumns(newR, names);
                                this.rows()[i] = newR;
                                return [2 /*return*/, Promise.resolve(this.translateFromJson(this.rows()[i], names))];
                            }
                        }
                        throw new Error("ArrayEntityDataProvider: Couldn't find row with id \"".concat(id, "\" in entity \"").concat(this.entity.key, "\" to update"));
                }
            });
        });
    };
    ArrayEntityDataProvider.prototype.delete = function (id) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var names, idMatches, i;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        names = _a.sent();
                        idMatches = this.idMatches(id, names);
                        for (i = 0; i < this.rows().length; i++) {
                            if (idMatches(this.rows()[i])) {
                                this.rows().splice(i, 1);
                                return [2 /*return*/, Promise.resolve()];
                            }
                        }
                        throw new Error("ArrayEntityDataProvider: Couldn't find row with id \"".concat(id, "\" in entity \"").concat(this.entity.key, "\" to delete"));
                }
            });
        });
    };
    ArrayEntityDataProvider.prototype.insert = function (data) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var names, j, idf, _a, _b, row;
            var e_6, _c;
            return tslib_1.__generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, this.init()];
                    case 1:
                        names = _d.sent();
                        j = this.translateToJson(data, names);
                        idf = this.entity.idMetadata.field;
                        if (!(idf instanceof CompoundIdField_js_1.CompoundIdField)) {
                            if (idf.valueConverter.fieldTypeInDb === 'autoincrement') {
                                j[idf.key] = 1;
                                try {
                                    for (_a = tslib_1.__values(this.rows()), _b = _a.next(); !_b.done; _b = _a.next()) {
                                        row = _b.value;
                                        if (row[idf.key] >= j[idf.key])
                                            j[idf.key] = row[idf.key] + 1;
                                    }
                                }
                                catch (e_6_1) { e_6 = { error: e_6_1 }; }
                                finally {
                                    try {
                                        if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                                    }
                                    finally { if (e_6) throw e_6.error; }
                                }
                            }
                            if (j[idf.key])
                                this.rows().forEach(function (i) {
                                    if (j[idf.key] == i[idf.key])
                                        throw Error('id already exists');
                                });
                        }
                        this.verifyThatRowHasAllNotNullColumns(j, names);
                        this.rows().push(j);
                        return [2 /*return*/, Promise.resolve(this.translateFromJson(j, names))];
                }
            });
        });
    };
    return ArrayEntityDataProvider;
}());
exports.ArrayEntityDataProvider = ArrayEntityDataProvider;
function pageArray(rows, options) {
    if (!options)
        return rows;
    if (!options.limit)
        return rows;
    var page = 1;
    if (options.page)
        page = options.page;
    if (page < 1)
        page = 1;
    var x = 0;
    return rows.filter(function (i) {
        x++;
        var max = page * options.limit;
        var min = max - options.limit;
        return x > min && x <= max;
    });
}
var FilterConsumerBridgeToObject = /** @class */ (function () {
    function FilterConsumerBridgeToObject(row, dbNames) {
        this.row = row;
        this.dbNames = dbNames;
        this.ok = true;
    }
    FilterConsumerBridgeToObject.prototype.databaseCustom = function (databaseCustom) {
        if (databaseCustom && databaseCustom.arrayFilter) {
            if (!databaseCustom.arrayFilter(this.row))
                this.ok = false;
        }
    };
    FilterConsumerBridgeToObject.prototype.custom = function (key, customItem) {
        throw new Error('Custom Filter should be translated before it gets here');
    };
    FilterConsumerBridgeToObject.prototype.or = function (orElements) {
        var e_7, _a;
        try {
            for (var orElements_1 = tslib_1.__values(orElements), orElements_1_1 = orElements_1.next(); !orElements_1_1.done; orElements_1_1 = orElements_1.next()) {
                var element = orElements_1_1.value;
                var filter = new FilterConsumerBridgeToObject(this.row, this.dbNames);
                element.__applyToConsumer(filter);
                if (filter.ok) {
                    return;
                }
            }
        }
        catch (e_7_1) { e_7 = { error: e_7_1 }; }
        finally {
            try {
                if (orElements_1_1 && !orElements_1_1.done && (_a = orElements_1.return)) _a.call(orElements_1);
            }
            finally { if (e_7) throw e_7.error; }
        }
        this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.not = function (element) {
        var filter = new FilterConsumerBridgeToObject(this.row, this.dbNames);
        element.__applyToConsumer(filter);
        if (filter.ok)
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isNull = function (col) {
        if (this.row[this.dbNames.$dbNameOf(col)] != null)
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isNotNull = function (col) {
        if (this.row[this.dbNames.$dbNameOf(col)] == null)
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isIn = function (col, val) {
        var e_8, _a;
        try {
            for (var val_1 = tslib_1.__values(val), val_1_1 = val_1.next(); !val_1_1.done; val_1_1 = val_1.next()) {
                var v = val_1_1.value;
                if (this.row[this.dbNames.$dbNameOf(col)] == col.valueConverter.toJson(v)) {
                    return;
                }
            }
        }
        catch (e_8_1) { e_8 = { error: e_8_1 }; }
        finally {
            try {
                if (val_1_1 && !val_1_1.done && (_a = val_1.return)) _a.call(val_1);
            }
            finally { if (e_8) throw e_8.error; }
        }
        this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isEqualTo = function (col, val) {
        if (this.row[this.dbNames.$dbNameOf(col)] != col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isDifferentFrom = function (col, val) {
        if (this.row[this.dbNames.$dbNameOf(col)] == col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isGreaterOrEqualTo = function (col, val) {
        if (this.row[this.dbNames.$dbNameOf(col)] < col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isGreaterThan = function (col, val) {
        if (this.row[this.dbNames.$dbNameOf(col)] <= col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isLessOrEqualTo = function (col, val) {
        if (this.row[this.dbNames.$dbNameOf(col)] > col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.isLessThan = function (col, val) {
        if (this.row[this.dbNames.$dbNameOf(col)] >= col.valueConverter.toJson(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.containsCaseInsensitive = function (col, val) {
        var v = this.row[this.dbNames.$dbNameOf(col)];
        if (!v) {
            this.ok = false;
            return;
        }
        var s = '' + v;
        if (val)
            val = col.valueConverter.toJson(val);
        if (val)
            val = val.toString().toLowerCase();
        if (s.toLowerCase().indexOf(val) < 0)
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.notContainsCaseInsensitive = function (col, val) {
        var v = this.row[this.dbNames.$dbNameOf(col)];
        if (!v) {
            this.ok = false;
            return;
        }
        var s = '' + v;
        if (val)
            val = col.valueConverter.toJson(val);
        if (val)
            val = val.toString().toLowerCase();
        if (s.toLowerCase().indexOf(val) >= 0)
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.startsWithCaseInsensitive = function (col, val) {
        var v = this.row[this.dbNames.$dbNameOf(col)];
        if (!v) {
            this.ok = false;
            return;
        }
        var s = '' + v;
        if (val)
            val = col.valueConverter.toJson(val);
        if (val)
            val = val.toString().toLowerCase();
        if (!s.toLowerCase().startsWith(val))
            this.ok = false;
    };
    FilterConsumerBridgeToObject.prototype.endsWithCaseInsensitive = function (col, val) {
        var v = this.row[this.dbNames.$dbNameOf(col)];
        if (!v) {
            this.ok = false;
            return;
        }
        var s = '' + v;
        if (val)
            val = col.valueConverter.toJson(val);
        if (val)
            val = val.toString().toLowerCase();
        if (!s.toLowerCase().endsWith(val))
            this.ok = false;
    };
    return FilterConsumerBridgeToObject;
}());
