// remult/src/data-providers/web-sql-data-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSqlDataProvider = void 0;
var tslib_1 = require("tslib");
var filter_consumer_bridge_to_sql_request_js_1 = require("../filter/filter-consumer-bridge-to-sql-request.js");
var RepositoryImplementation_js_1 = require("../remult3/RepositoryImplementation.js");
var sql_database_js_1 = require("./sql-database.js");
//SqlDatabase.LogToConsole = true;
var WebSqlDataProvider = /** @class */ (function () {
    function WebSqlDataProvider(databaseName, databaseSize) {
        if (databaseSize === void 0) { databaseSize = 2 * 1024 * 1024; }
        this.databaseName = databaseName;
        //@ts-ignore
        this.db = window.openDatabase(databaseName, '1.0', databaseName, databaseSize);
    }
    WebSqlDataProvider.prototype.end = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/];
        }); });
    };
    WebSqlDataProvider.prototype.getLimitSqlSyntax = function (limit, offset) {
        return ' limit ' + limit + ' offset ' + offset;
    };
    WebSqlDataProvider.prototype.entityIsUsedForTheFirstTime = function (entity) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.createTable(entity)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    WebSqlDataProvider.prototype.ensureSchema = function (entities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var entities_1, entities_1_1, entity, e_1_1;
            var e_1, _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 5, 6, 7]);
                        entities_1 = tslib_1.__values(entities), entities_1_1 = entities_1.next();
                        _b.label = 1;
                    case 1:
                        if (!!entities_1_1.done) return [3 /*break*/, 4];
                        entity = entities_1_1.value;
                        return [4 /*yield*/, this.createTable(entity)];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3:
                        entities_1_1 = entities_1.next();
                        return [3 /*break*/, 1];
                    case 4: return [3 /*break*/, 7];
                    case 5:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 7];
                    case 6:
                        try {
                            if (entities_1_1 && !entities_1_1.done && (_a = entities_1.return)) _a.call(entities_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    WebSqlDataProvider.prototype.dropTable = function (entity) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var e, sql;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, filter_consumer_bridge_to_sql_request_js_1.dbNamesOf)(entity)];
                    case 1:
                        e = _a.sent();
                        sql = 'drop  table if exists ' + e.$entityName;
                        if (sql_database_js_1.SqlDatabase.LogToConsole)
                            console.info(sql);
                        return [4 /*yield*/, this.createCommand().execute(sql)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    WebSqlDataProvider.prototype.createTable = function (entity) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var result, e, _a, _b, x, sql;
            var e_2, _c;
            return tslib_1.__generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        result = '';
                        return [4 /*yield*/, (0, filter_consumer_bridge_to_sql_request_js_1.dbNamesOf)(entity)];
                    case 1:
                        e = _d.sent();
                        try {
                            for (_a = tslib_1.__values(entity.fields), _b = _a.next(); !_b.done; _b = _a.next()) {
                                x = _b.value;
                                if (!(0, filter_consumer_bridge_to_sql_request_js_1.isDbReadonly)(x, e) || (0, RepositoryImplementation_js_1.isAutoIncrement)(x)) {
                                    if (result.length != 0)
                                        result += ',';
                                    result += '\r\n  ';
                                    if ((0, RepositoryImplementation_js_1.isAutoIncrement)(x)) {
                                        if (x.key != entity.idMetadata.field.key)
                                            throw 'in web sql, autoincrement is only allowed for primary key';
                                        result += e.$dbNameOf(x) + ' integer primary key autoincrement';
                                    }
                                    else {
                                        result += this.addColumnSqlSyntax(x, e.$dbNameOf(x));
                                        if (x.key == entity.idMetadata.field.key) {
                                            result += ' primary key';
                                        }
                                    }
                                }
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                        sql = 'create table if not exists ' + e.$entityName + ' (' + result + '\r\n)';
                        if (sql_database_js_1.SqlDatabase.LogToConsole)
                            console.log(sql);
                        return [4 /*yield*/, this.createCommand().execute(sql)];
                    case 2:
                        _d.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    WebSqlDataProvider.prototype.createCommand = function () {
        return new WebSqlBridgeToSQLCommand(this.db);
    };
    WebSqlDataProvider.prototype.transaction = function (action) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                throw new Error('Method not implemented.');
            });
        });
    };
    WebSqlDataProvider.prototype.addColumnSqlSyntax = function (x, dbName) {
        var result = dbName;
        var nullNumber = x.allowNull ? '' : ' default 0 not null';
        if (x.valueType == Date)
            result += ' integer';
        else if (x.valueType == Boolean)
            result += ' integer ' + nullNumber;
        else if (x.valueType == Number) {
            if (!x.valueConverter.fieldTypeInDb)
                result += ' real ' + nullNumber;
            else
                result += ' ' + x.valueConverter.fieldTypeInDb + ' ' + nullNumber;
        }
        else
            result += ' text' + (x.allowNull ? ' ' : " default '' not null ");
        return result;
    };
    WebSqlDataProvider.prototype.toString = function () {
        return 'WebSqlDataProvider';
    };
    return WebSqlDataProvider;
}());
exports.WebSqlDataProvider = WebSqlDataProvider;
var WebSqlBridgeToSQLCommand = /** @class */ (function () {
    //@ts-ignore
    function WebSqlBridgeToSQLCommand(source) {
        this.source = source;
        this.values = [];
    }
    WebSqlBridgeToSQLCommand.prototype.addParameterAndReturnSqlToken = function (val) {
        return this.param(val);
    };
    WebSqlBridgeToSQLCommand.prototype.param = function (val) {
        this.values.push(val);
        return '~' + this.values.length + '~';
    };
    WebSqlBridgeToSQLCommand.prototype.execute = function (sql) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            return _this.source.transaction(function (t) {
                var s = sql;
                var v = [];
                var m = s.match(/~\d+~/g);
                if (m != null)
                    m.forEach(function (mr) {
                        s = s.replace(mr, '?');
                        v.push(_this.values[Number.parseInt(mr.substring(1, mr.length - 1)) - 1]);
                    });
                t.executeSql(s, v, function (t1, r) { return resolve(new WebSqlBridgeToSQLQueryResult(r)); }, function (t2, err) {
                    reject(err.message);
                    return false;
                });
            });
        });
    };
    return WebSqlBridgeToSQLCommand;
}());
var WebSqlBridgeToSQLQueryResult = /** @class */ (function () {
    //@ts-ignore
    function WebSqlBridgeToSQLQueryResult(r) {
        this.r = r;
        this.rows = [];
        for (var i = 0; i < r.rows.length; i++) {
            this.rows.push(r.rows.item(i));
        }
    }
    WebSqlBridgeToSQLQueryResult.prototype.getColumnKeyInResultForIndexInSelect = function (index) {
        if (this.rows.length == 0)
            throw Error('No rows');
        var i = 0;
        for (var m in this.rows[0]) {
            if (i++ == index)
                return m;
        }
        throw Error('index not found');
    };
    return WebSqlBridgeToSQLQueryResult;
}());
