// remult/migrations/compare-migration-snapshots.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.emptySnapshot = exports.compareMigrationSnapshot = void 0;
var tslib_1 = require("tslib");
var index_js_1 = require("../index.js");
var filter_consumer_bridge_to_sql_request_js_1 = require("../src/filter/filter-consumer-bridge-to-sql-request.js");
function compareMigrationSnapshot(_a) {
    var entities = _a.entities, snapshot = _a.snapshot, migrationBuilder = _a.migrationBuilder, reporter = _a.reporter;
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var processedEntities, entities_1, entities_1_1, entity, meta, e, entitySnapshot, createColumns, processedColumns, _b, _c, field, columnDbName, column, e_1_1, _d, _e, columnDbName, e_2_1, e_3_1, _f, _g, entityDbName, e_4_1;
        var e_3, _h, e_1, _j, e_2, _k, e_4, _l;
        return tslib_1.__generator(this, function (_m) {
            switch (_m.label) {
                case 0:
                    snapshot = JSON.parse(JSON.stringify(snapshot));
                    processedEntities = new Set();
                    _m.label = 1;
                case 1:
                    _m.trys.push([1, 23, 24, 25]);
                    entities_1 = tslib_1.__values(entities), entities_1_1 = entities_1.next();
                    _m.label = 2;
                case 2:
                    if (!!entities_1_1.done) return [3 /*break*/, 22];
                    entity = entities_1_1.value;
                    meta = (0, index_js_1.repo)(entity).metadata;
                    return [4 /*yield*/, (0, filter_consumer_bridge_to_sql_request_js_1.dbNamesOf)(meta, function (x) { return x; })];
                case 3:
                    e = _m.sent();
                    processedEntities.add(e.$entityName);
                    if (!(0, filter_consumer_bridge_to_sql_request_js_1.shouldCreateEntity)(meta, e)) return [3 /*break*/, 21];
                    entitySnapshot = snapshot.entities[e.$entityName];
                    createColumns = true;
                    if (!!entitySnapshot) return [3 /*break*/, 5];
                    createColumns = false;
                    reporter('create table ' + e.$entityName);
                    return [4 /*yield*/, migrationBuilder.createTable(meta)];
                case 4:
                    _m.sent();
                    entitySnapshot = snapshot.entities[e.$entityName] = {
                        key: meta.key,
                        className: meta.entityType.name,
                        columns: {},
                    };
                    _m.label = 5;
                case 5:
                    processedColumns = new Set();
                    _m.label = 6;
                case 6:
                    _m.trys.push([6, 12, 13, 14]);
                    _b = (e_1 = void 0, tslib_1.__values(meta.fields)), _c = _b.next();
                    _m.label = 7;
                case 7:
                    if (!!_c.done) return [3 /*break*/, 11];
                    field = _c.value;
                    if (!!(0, filter_consumer_bridge_to_sql_request_js_1.shouldNotCreateField)(field, e)) return [3 /*break*/, 10];
                    columnDbName = e.$dbNameOf(field);
                    processedColumns.add(columnDbName);
                    column = entitySnapshot.columns[columnDbName];
                    if (!!column) return [3 /*break*/, 10];
                    if (!createColumns) return [3 /*break*/, 9];
                    reporter('add column ' + e.$entityName + '.' + columnDbName);
                    return [4 /*yield*/, migrationBuilder.addColumn(meta, field)];
                case 8:
                    _m.sent();
                    _m.label = 9;
                case 9:
                    entitySnapshot.columns[columnDbName] = { key: field.key };
                    _m.label = 10;
                case 10:
                    _c = _b.next();
                    return [3 /*break*/, 7];
                case 11: return [3 /*break*/, 14];
                case 12:
                    e_1_1 = _m.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 14];
                case 13:
                    try {
                        if (_c && !_c.done && (_j = _b.return)) _j.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7 /*endfinally*/];
                case 14:
                    _m.trys.push([14, 19, 20, 21]);
                    _d = (e_2 = void 0, tslib_1.__values(Object.keys(entitySnapshot.columns))), _e = _d.next();
                    _m.label = 15;
                case 15:
                    if (!!_e.done) return [3 /*break*/, 18];
                    columnDbName = _e.value;
                    if (!!processedColumns.has(columnDbName)) return [3 /*break*/, 17];
                    reporter('remove column ' + e.$entityName + '.' + columnDbName);
                    return [4 /*yield*/, migrationBuilder.removeColumn(e.$entityName, columnDbName)];
                case 16:
                    _m.sent();
                    delete entitySnapshot.columns[columnDbName];
                    _m.label = 17;
                case 17:
                    _e = _d.next();
                    return [3 /*break*/, 15];
                case 18: return [3 /*break*/, 21];
                case 19:
                    e_2_1 = _m.sent();
                    e_2 = { error: e_2_1 };
                    return [3 /*break*/, 21];
                case 20:
                    try {
                        if (_e && !_e.done && (_k = _d.return)) _k.call(_d);
                    }
                    finally { if (e_2) throw e_2.error; }
                    return [7 /*endfinally*/];
                case 21:
                    entities_1_1 = entities_1.next();
                    return [3 /*break*/, 2];
                case 22: return [3 /*break*/, 25];
                case 23:
                    e_3_1 = _m.sent();
                    e_3 = { error: e_3_1 };
                    return [3 /*break*/, 25];
                case 24:
                    try {
                        if (entities_1_1 && !entities_1_1.done && (_h = entities_1.return)) _h.call(entities_1);
                    }
                    finally { if (e_3) throw e_3.error; }
                    return [7 /*endfinally*/];
                case 25:
                    _m.trys.push([25, 30, 31, 32]);
                    _f = tslib_1.__values(Object.keys(snapshot.entities)), _g = _f.next();
                    _m.label = 26;
                case 26:
                    if (!!_g.done) return [3 /*break*/, 29];
                    entityDbName = _g.value;
                    if (!!processedEntities.has(entityDbName)) return [3 /*break*/, 28];
                    reporter('remove table ' + entityDbName);
                    return [4 /*yield*/, migrationBuilder.removeTable(entityDbName)];
                case 27:
                    _m.sent();
                    delete snapshot.entities[entityDbName];
                    _m.label = 28;
                case 28:
                    _g = _f.next();
                    return [3 /*break*/, 26];
                case 29: return [3 /*break*/, 32];
                case 30:
                    e_4_1 = _m.sent();
                    e_4 = { error: e_4_1 };
                    return [3 /*break*/, 32];
                case 31:
                    try {
                        if (_g && !_g.done && (_l = _f.return)) _l.call(_f);
                    }
                    finally { if (e_4) throw e_4.error; }
                    return [7 /*endfinally*/];
                case 32: return [2 /*return*/, snapshot];
            }
        });
    });
}
exports.compareMigrationSnapshot = compareMigrationSnapshot;
function emptySnapshot() {
    return {
        version: 1,
        entities: {},
    };
}
exports.emptySnapshot = emptySnapshot;
